[28기 강희원][Add] 추가 기능 구현중

KH2Wone opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x] 기능 추가
- [ ] 리뷰 반영
- [x] 리팩토링
- [x] 버그 수정
- [x] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
- 로그인, 메인 페이지 구현 (완료)
- main 페이지에서 댓글 기능 구현 (완료)
- 추가 기능 구현
 - [x] Mission 5 | ID, PW validation
 - [ ] Mission 6 | 댓글 좋아요 / 삭제 기능
 - [ ] Mission 7 | 아이디 검색 기능
 - [ ] Mission 8 | nav 프로필 사진 클릭 시 메뉴 박스 생성
 - [x] Mission 9 | 반응형 구현

<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
1. 로그인 페이지 작업
- 로그인 페이지 레이아웃
- 로그인 버튼 활성화 및 비활성화
- 유효성 검사
 - 비밀번호 5글자 이상, 아이디 '@' 문자 포함이자 문자열 10개 이상 (휴대폰 번호나 아이디일 수 있으므로)

2. main 페이지 구현
-  main 페이지 레이아웃
- 댓글 게시 기능
- 추가 기능 구현중


<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
- `addEventListener`를 통해 로그인 버튼을 사용자의 입력값에 따라 `keyup` 등의 이벤트를 줄 수 있다는 것을 새로 알게되었습니다.
- `display: flex`를 다양하게 적용해볼 수 있는 기회가 되었으나 아직 익숙하지 않습니다.
- CSS 속성의 우선순위에 따라 리팩토링 해볼 수 있었습니다.
- `position: fixed`는 부모에게 영향을 받지 않고 브라우저에서 독단적인 위치를 사용합니다.
그러나 위스타그램의 경우 `aside`의 위치가 고정되면서 특정 위치에 PC 사이즈가 늘었다 줄어도 항상 그 자리에 있어야 하기에 `transform: translate(643px)`을 적용하여 위치를 고정시켜줄 수 있었습니다.
인스타그램의 경우 JS로 브라우저의 크기에 따라 `left`의 위치가 유동적으로 변하게 했으나 저는 `transform` 속성으로 고정시켜두어 보았습니다. `left`가 유동적으로 변하게 하는 로직을 구현해봐야겠습니다.
- `flex` 속성, `flex-grow` 속성을 더 공부해봐야합니다.
- `object-fit`을 사용할 때 해당 속성은 적용이 되지만 `object-position`을 사용하려면 `width`, `height`의 값이 지정되어 있어야 합니다.
- block 요소에게는 `width`나 `display: block`을 주는 것은 불필요하므로 생략해야 합니다.
- 인스타그램의 경우 댓글 영역을 `textarea`로 만들고 댓글의 길이(양)에 따라 스크롤이나 `overflow: hidden`이 아닌 유동적으로 `height`가 늘어나 모든 내용이 보이도록 했습니다.
해당 내용을 구현해볼 필요가 있겠습니다.
- `max-width`를 잘 활용해서 쉽게 반응형을 구현할 수 있습니다. 전에는 잘 사용하지 않았지만 이제는 자주 적용해볼 수 있겠습니다.
- 핸드폰에서 사용하는 어플인 만큼 사용자의 손가락이 크더라도 원하는 곳을 터치할 수 있어야 합니다. 그래서 인스타그램의 경우 `padding`을 줘서 영역을 차지해 클릭할 수 있는 범위를 넓히고 그로 인해 부모 레이아웃을 벗어날 경우 `margin`으로 마이너스 값을 주어 위치하게 했습니다.
- `flex`를 더 많이 써봐야겠습니다.

<br />

## :: 기타 질문 및 특이 사항
- `querySelector`보다 `getElementById`의 효율이 더 좋다고 해서 사용하려고 지향하는데 login.html에서 `class` 및 `id`가 너무 남발된건 아닌지 가급적 `class`와 `id`의 사용을 자제하는게 좋은지 궁금합니다.

claire-euni commented on students/14th/donghaKim/style/login.css:
```diff
@@ -0,0 +1,271 @@
+* {
+    margin: 0;
+    padding: 0;
+    box-sizing: border-box;
+}
+
+body {}
```
이 부분은 common.css에서 관리해도 되겠죠?

dongha1992 replied:
확인했습니다.

claire-euni commented on students/14th/donghaKim/style/login.css:
```diff
@@ -0,0 +1,271 @@
+.main-wrap {
+    display: flex;
+    justify-content: center;
+    width: 100%;
```
필요할까요~?

dongha1992 replied:
가끔 헷갈릴 때마다 이것 저것 주다가 저렇게 되는데 그래서 그런 것 같습니다.

claire-euni commented on students/14th/donghaKim/style/login.css:
```diff
@@ -0,0 +1,271 @@
+}
+
+.right-wrap {
+    width: 400px;
```
부모에서 이미 width값을 주었기 때문에 중복되는 속성이라 삭제해도 무방합니다 :)

dongha1992 replied:
width의 부모자식 관계는 늘 헷갈리네요..

claire-euni commented on students/14th/donghaKim/style/login.css:
```diff
@@ -0,0 +1,271 @@
+}
+
+.divided::before {
+    content: "";
+    position: absolute;
+    width: 100px;
+    height: 1px;
+    color: black;
+    background-color: rgba(0, 0, 0, 0.3);
+    left: 0;
+    bottom: 45px;
+}
+
+.divided::after {
+    content: "";
+    position: absolute;
+    width: 100px;
+    height: 1px;
+    background-color: rgba(0, 0, 0, 0.3);
+    color: black;
+    right: 0px;
+    bottom: 45px;
+}
```
반복되는 속성이라면 한번에 묶어서 써줄 수 있겠죠?  그리고 before와 after가 필요한 상황인가요?

dongha1992 replied:
멋있어 보여 사용해봤습니다! div으로 다시 만들었습니다.

claire-euni commented on students/14th/donghaKim/style/login.css:
```diff
@@ -0,0 +1,271 @@
+    height: 20px;
+}
+
+.facebook-login a {
+    outline: none;
+    text-decoration: none;
```
이 부분은 주로 common.css에서 a 태그에 적용되는 속성들입니다. 모든 a 태그에 대해서 필요한 속성이라면 common.css에 넣는것을 고려해보세요 :)

dongha1992 replied:
뭔가 아직 어떤 것이 공통적으로 필요한 건지 잡히지기 않아서 공통 css는 생각이 잘 안 나긴 하는데 노력하겠습니다!

claire-euni commented on students/14th/donghaKim/style/login.css:
```diff
@@ -0,0 +1,271 @@
+    padding-bottom: 10px;
+}
+
+.right-app {}
```
사용되지 않는 셀렉터는 삭제해주세요~

dongha1992 replied:
네!

claire-euni commented on students/14th/donghaKim/style/login.css:
```diff
@@ -0,0 +1,271 @@
+}
+
+.download-app .app-btn button {
+    width: 155px;
+    background-color: black;
+    height: 43px;
+    border-radius: 10px;
+    outline: none;
+    border: none;
```
관련도 있는 css속성끼리 묶어서 순서대로 작성해주시는게 좋습니다! 예를 들어 display / width,height / padding,margin / border / position / font 등으로 작성해주세요 :)

dongha1992 replied:
네 알겠습니다!

claire-euni commented on students/14th/donghaKim/style/login.css:
```diff
@@ -0,0 +1,271 @@
+    left: 55%;
+}
+
+.show-success {}
```
css에 대한 팁을 드리자면 개발자도구를 켜서 체크박스에서 지워보면서 중복되거나 불필요한 속성들을 확인하는것이 좋습니다 :)

dongha1992 replied:
css를 다 하고 나면 다시 보는데 엄청난 에너지가 필요해서 다시 잘 보지 않게 되는 거 같습니다. 하지만 앞으로 보는 노력을 하겠습니다.

claire-euni commented on students/14th/donghaKim/js/main.js:
```diff
@@ -0,0 +1,353 @@
+navLogout.addEventListener("click", logoutHandler)
+
+
+const listsContainer = document.querySelector(".feed-user-comment-list-wrap")
+const feedUserForm = document.querySelector(".feed-user-form-js")
+const feedUserInput = document.querySelector(".feed-user-comment-input")
+const showRefList = document.querySelector(".refList")
+const contentMore = document.querySelector(".feed-user-content-more")
+const userContent = document.querySelector(".feed-user-content")
+const commentList = document.querySelector(".feed-user-comment-count")
+const commentCountSpan = document.querySelector(".comment-count")
+const commentShowBtn = document.querySelector(".comment-show-btn")
+const storyListsWrap = document.querySelector(".story-lists-wrap")
+const storyDeleteBtn = document.querySelector(".story-delete")
+const storyWrapTarget = document.querySelector(".story-add-wrap")
+const searchForm = document.querySelector(".header-search-form")
+const formForBorder = document.querySelector(".form-wrap")
+const searchUserInput = document.querySelector(".header-search-input")
+const card = document.querySelector(".card")
+const insertBefore = document.querySelector(".feed-header-wrap")
+const insertAfter = document.querySelector(".bottom")
+const feedRender = document.querySelector(".feed-render")
+const refListWrap = document.querySelector(".refList-wrap")
+
+let tempStr = ""
+let submited = false
+let showContentFlag = false
+let showCommentFlag = true;
```
전역 변수들은 같이 상단에 위치하게 해주세요~

dongha1992 replied:
nav와 main page에서 쓰이는 기능을 분리하기 위해 전역변수를 나눴는데 나쁜 습관이었군요., 수정했습니다.

claire-euni commented on students/14th/donghaKim/js/main.js:
```diff
@@ -0,0 +1,353 @@
+    }
+})
+
+function saveLists() {
+    localStorage.setItem(LOCAL_KEY, JSON.stringify(lists))
+    localStorage.setItem(STORY_KEY, JSON.stringify(userArr))
+    localStorage.setItem(SEARCH_KEY, JSON.stringify(serachArr))
+}
```
localStorage에 저장해서 JSON도 사용하시고 와우 👍

dongha1992 replied:
감사합니다

claire-euni commented on students/14th/donghaKim/js/main.js:
```diff
@@ -0,0 +1,353 @@
+let STORY_KEY = "story-key"
+let SEARCH_KEY = "search-key"
+let lists = JSON.parse(localStorage.getItem(LOCAL_KEY)) || []
+    // let userArr = JSON.parse(localStorage.getItem(STORY_KEY)) || []
```
주석부분은 삭제해주세요~

dongha1992 replied:
삭제했습니다.

claire-euni commented on students/14th/donghaKim/js/main.js:
```diff
@@ -0,0 +1,353 @@
+    }
+}
+
+function heartCount() {
+    if (click) {
+        cnt++
+    } else {
+        cnt--
+    }
+    let preHeart = document.querySelector(".feed-like").innerHTML
+    preHeart = `${cnt}`
+    document.querySelector(".feed-like").innerHTML = preHeart
+    console.log(preHeart)
+}
```
아래와 같이 써줄수도 있겠죠? 
```js
function heartCount() {
  click ? cnt++ : cent--;
  document.querySelector(".feed-like").innerHTML = `${cnt}`;
 }
```

dongha1992 replied:
써줄 수 있겠지만 아직 제 머리로는 무리인 거 같습니다. 수정했습니다!

claire-euni commented on students/14th/donghaKim/js/main.js:
```diff
@@ -0,0 +1,353 @@
+}
+heartCount()
+
+function storyAddHandler() {
+    storyAddDiv.classList.toggle("show-profile")
+}
+
+function navProdilehandler() {
+    toggledProfile.classList.toggle("show-profile")
+}
```
toggle 잘 구현하셨습니다! 🙌

dongha1992 replied:
감사합니다!

claire-euni commented on students/14th/donghaKim/js/main.js:
```diff
@@ -0,0 +1,353 @@
+
+function createStoryObj(userId, url) {
+    return {
+        id: new Date().getTime().toString(),
```
Date() 사용해서 고유 id값을 만들었군요!

dongha1992 replied:
칭찬을 들으니 더욱 코딩 할 맛이나네요!

claire-euni commented on students/14th/donghaKim/js/main.js:
```diff
@@ -0,0 +1,353 @@
+heartWrap.addEventListener("click", navHeartHandler)
+feedWrap.addEventListener("click", feedHeartHandler)
+heartWrap.addEventListener("click", storyAddHandler)
+storyAddDiv.addEventListener("submit", function(e) {
+    e.preventDefault()
+    let userId = storyAddName.value
+    let userUrl = storyAddImg.value
+    storyAddName.value = ""
+    storyAddImg.value = ""
+    let story = createStoryObj(userId, userUrl)
+    userArr.push(story)
+    renderStory()
```
이 부분도 로그인 부분 리뷰 참고해보시면 좋을것 같습니다~

dongha1992 replied:
넵!

claire-euni commented on students/14th/donghaKim/js/main.js:
```diff
@@ -0,0 +1,353 @@
+}
+
+
+function clearEelmenet(container) {
```
함수명에 오타가 있습니다 😎

dongha1992 replied:
눈 아프셨을 텐데 감사합니다!

claire-euni commented on students/14th/donghaKim/js/main.js:
```diff
@@ -0,0 +1,353 @@
+}]
+
+function navHeartHandler(e) {
+    const heart = e.target
+    if (heart.classList.contains("i-default")) {
+        navHeartAfter.style.display = "block"
+    } else if (heart.classList.contains("full")) {
+        navHeartAfter.style.display = "none"
+    }
+}
+
+function feedHeartHandler(e) {
+    const heart = e.target
```
두 heart가 같은것을 의미하는게 아니라면 각자 구분될 수 있게 변수명을 고민해보세요~

dongha1992 replied:
급하게 수정하느라 대충 적었는데 고치지 못했습니다. 다음부터는 클래스 하나 하나 소중히 적겠습니다.

claire-euni commented on students/14th/donghaKim/js/main.js:
```diff
@@ -0,0 +1,353 @@
+    saveLists()
+}
+
+function renderStory() {
+    clearEelmenet(storyListsWrap)
+    userArr.forEach(story => {
+        const storyElement = document.createElement("li")
+        storyElement.classList.add("feed-user-comment-list")
+        storyElement.dataset.listId = story.id
+        storyElement.innerHTML += ` <img src="${story.bg}" alt="" class="circle">
+                                    <img src="${story.src}" alt="" class="ciicle-avatar">
+                                    <span class="avatar-name">${story.userId}</span>
+                                 </li>`
+        storyListsWrap.appendChild(storyElement)
+    })
+    saveLists()
+}
```
이 부분은 스토리를 등록하는건가요? 화면에서는 보이지 않아서요

dongha1992 replied:
<img width="516" alt="스크린샷 2020-10-30 오전 11 51 54" src="https://user-images.githubusercontent.com/58624851/97654479-505b8300-1aa6-11eb-94ed-c9102f043faa.png">

nav에 heart를 누르면 나오는데 뭔가 UI상 표시를 하고 싶었는데 그럴 깜냥이 되지 않아 일단 그냥 뒀습니다.

claire-euni commented on students/14th/donghaKim/js/main.js:
```diff
@@ -0,0 +1,353 @@
+    saveLists()
+}
+
+function showList(userInputCommment, submited) {
+    if (userInputCommment[1]) {
+        if (!submited) {
+            let targetedUser = []
+            for (let i = 0; i < userArr[0].userId.length; i++) {
+                let targetUserId = userArr[i].userId
+                let newInput = userInputCommment.slice(1, userInputCommment.length)
+                if (targetUserId.indexOf(newInput) > -1) {
+                    targetedUser.push(userArr[i])
+                }
+            }
+            targetedUser.forEach(user => {
+                const div = document.createElement("div")
+                div.classList.add("refList-wrap")
+                div.innerHTML += `             
+                                  <div class="refList-avatar">
+                                      <img src="img/main/story_icon.png" alt="" class="refList-bg">
+                                      <img src=${user.src} alt="" class="refList-img">
+                                       </div>
+                                        <div class="ref-name">
+                                           <span class="ref-id-span">${user.userId}</span>
+                                   </div>
+                             </div >
+                        `
+                showRefList.appendChild(div)
+            })
+        }
+    }
+}
```
이 부분은 리뷰 진행중

dongha1992 replied:
앗 운정님 여기 함수는 검색 기능인데 지금 저도 수정 중입니다!

claire-euni reviewed (commented):
동하님 첫 PR 잘하셨습니다! 많은 기능 구현해주셨네요! 리뷰반영해서 리팩토링 진행해주세요 😄

dongha1992 commented on students/14th/donghaKim/login.html:
```diff
@@ -0,0 +1,67 @@
+<!DOCTYPE html>
+<html>
+
+<head>
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width">
+    <title>login</title>
+    <link href="style/common.css" rel="stylesheet" type="text/css" />
+    <link href="style/login.css" rel="stylesheet" type="text/css" />
+</head>
+
```
<img width="516" alt="스크린샷 2020-10-30 오후 4 03 01" src="https://user-images.githubusercontent.com/58624851/97670362-6464ac00-1ac9-11eb-85b4-a7bd34b6d859.png">

엔터 간격이 이런 것을 말씀하시는 건가요??

dongha1992 commented on students/14th/donghaKim/login.html:
```diff
@@ -0,0 +1,67 @@
+                            <input type="password" class="password" placeholder="비밀번호를 입력해주세요.">
+                            <button class="login-btn">로그인</button>
+                        </form>
+                        <div class="divided">또는</div>
```
네 수정했습니다!

dongha1992 commented on students/14th/donghaKim/login.html:
```diff
@@ -0,0 +1,67 @@
+                        </form>
+                        <div class="divided">또는</div>
+                        <div class="facebook-login">
+                            <div class="col">
+                                <img src="img/login/facebook.jpeg" alt="">
+                                <a href="" class="des">페이스북으로 로그인</a>
+                            </div>
+                            <a href="" class="find">비밀 번호를 잊으셨나요?</a>
```
네 알겠습니다! 쓸지 안 쓸지 모르겠어서 일단 대충 적었는데 다음부터는 하나하나 잘 적겠습니다

dongha1992 commented on students/14th/donghaKim/js/login.js:
```diff
@@ -0,0 +1,45 @@
+
+
+form.addEventListener("keyup", function(e) {
+    let userPassword = passwordInput.value
+    userPassword.length > 6 ? loginBtn.style.backgroundColor = "#4e5dfc" : loginBtn.style.backgroundColor = ""
```
삼항연산자 쓸 때마다 헷갈리는데 연습해보겠습니다!

dongha1992 commented on students/14th/donghaKim/js/login.js:
```diff
@@ -0,0 +1,45 @@
+function validation() {
+    if (loginObj[0].userId === "immerfernweh_" && loginObj[0].userPassword === "1234") {
+        loginSuccess.style.display = "block"
+        timerId = setTimeout(function() { window.location = "main.html" }, 1000)
```
감사합니다~

dongha1992 commented on students/14th/donghaKim/js/login.js:
```diff
@@ -0,0 +1,45 @@
+    userPassword.length > 6 ? loginBtn.style.backgroundColor = "#4e5dfc" : loginBtn.style.backgroundColor = ""
+})
+
+form.addEventListener("submit", function(e) {
+    e.preventDefault()
+    let userId = IdInput.value
+    let userPassword = passwordInput.value
+    IdInput.value = ""
+    passwordInput.value = ""
+    let userInfo = createObj(userId, userPassword)
+    if (loginObj.length > 1) {
+        return
+    }
+    loginObj.push(userInfo)
+    validation()
+})
+
+function createObj(userId, userPassword) {
+    return {
+        userId,
+        userPassword
+    }
+}
```
1. 객체로 아이디를 저장해서 벨리데이션 할 때 데이터를 가져오려고 했는데 그런 방식으로 하지 않아 지워도 무방합니다! 지웠습니다. 그리고 무언가 회원가입? 이런 것도 만들고 싶어서 해놨는데 힘들 거 같아서 포기하면서 필요 없는 객체가 되었습니다. 

2. 아하... 저는 submit을 할 때면 필수적으로 preventDefault를 해서 새로고침을 막아야 하는 줄 알았습니다!

KH2Wone committed:
[Add] Login 페이지 완성 및 main 페이지 작업중
	- login 페이지 레이아웃 및 로그인 버튼 활성화&비활성화
	- main 페이지 레이아웃 완성
	- main 페이지 css 작업중
	- main 페이지 댓글 기능 작업 전

KH2Wone committed:
[Add] main CSS aside, users story section
- main 큰 레이아웃 완성
- 회원 추천 aside section CSS 구현
- 회원 스토리 리스트 박스 CSS 구현
- 회원 피드 article 구현중

KH2Wone committed:
Add: 메인 레이아웃 CSS 구현
- 이미지 랜덤하게 임의 적용
- input, textarea CSS  초기화
- CSS 속성 순서 리팩토링
- 메인 레이아웃 CSS 전체 구현 완료

KH2Wone committed:
Add: main 댓글 기능 구현
- Enter 키, 게시 버튼 클릭하면 댓글 생성
- 댓글 입력값 유효성
입력값이 들어와 있는 상태에서 Enter or 버튼 클릭했을 시에만 댓글 게시, 입력 없는 상태일 시 alert 경고창

KH2Wone committed:
[Fix] login, main js 리팩토링
- login, main js 리팩토링
- main header 사이즈 조정

KH2Wone committed:
[Fix] CSS inline 속성 요소 자동 여백 상쇄하여 직접 여백 적용, 불필요한 CSS 제거
